lib/AST/ModuleLoader.cpp

```diff
@@ -54,4 +59,73 @@ DependencyTracker::getClangCollector() {
   return clangCollector;
 }
 
+static bool findOverlayFilesInDirectory(SourceLoc diagLoc, StringRef path,
+                                        ModuleDecl *module,
+                                        DependencyTracker * const tracker) {
+  using namespace llvm::sys;
+  using namespace file_types;
+
+  ASTContext &ctx = module->getASTContext();
+  auto fs = ctx.SourceMgr.getFileSystem();
+
+  std::error_code error;
+  for (auto dir = fs->dir_begin(path, error);
+       !error && dir != llvm::vfs::directory_iterator();
+       dir.increment(error)) {
+    StringRef file = dir->path();
+    if (lookupTypeForExtension(path::extension(file)) != TY_SwiftOverlayFile)
+      continue;
+
+    module->addCrossImportOverlayFile(file);
+
+    // FIXME: Better to add it only if we load it.
+    if (tracker)
+      tracker->addDependency(file, module->isSystemModule());
+  }
+
+  if (error && error != std::errc::no_such_file_or_directory) {
+    ctx.Diags.diagnose(diagLoc, diag::cannot_list_swiftcrossimport_dir,
+                       module->getName(), error.message(), path);
+  }
+  return !error;
+}
+
+void ModuleLoader::findOverlayFiles(SourceLoc diagLoc, ModuleDecl *module,
+                                    FileUnit *file) {
+  using namespace llvm::sys;
+  using namespace file_types;
+
+  auto &langOpts = module->getASTContext().LangOpts;
+
+  SmallString<64> dirPath{file->getModuleDefiningPath()};
+
+  if (dirPath.empty())
+    return;
+  path::remove_filename(dirPath);
+
+  // <parent-dir>/<module-name>.swiftcrossimport
```

Comment:
I can't quite remember -- what was the reason behind us deciding to support both `swiftcrossimport` as a file and as a directory? Avoid more disk usage in case all the architecture-specific files are the same?

Reply:
The comments may not be clear enough here. What's happening is that:

1. `getModuleDefiningPath()` returns the path of the `module.map` file or `Foo.swiftmodule` directory.
2. We strip away the last component to get the directory that file is in.
3. We append `Foo.swiftcrossimport` to it, then search that directory.
4. We append the target's module triple to it, then search that directory.
5. If there's a target variant (for Catalyst), we remove the target's module triple and append the target variant's module triple, then search that directory.

If this isn't being conveyed well, how would you clarify it?

Reply:
Ah, gotcha'. Those are all directory searches. I just misread the first one.

I think the code is 👍 but I'm wondering if there should be extra context. Like, when would we use the top-level directory over the architecture-specific directory? Is it a question of not unnecessarily duplicating files?

Also, having a line like `// E.g. for Catalyst` would be useful immediately above the `// <parent-dir>/<module-name>.swiftcrossimport/<targetvariant-module-triple>`.

Reply:
Honestly, I don't know if we'll end up needing the target-specific directories; I added them because (a) it's easy to just do it now and (b) if we do need them, we will inevitably discover this only at the eleventh hour when there might be pressure to insert a hack instead.

Reply:
Expanded comments.